File: Solutions/110-Balanced Binary Tree/110-Balanced Binary Tree.cs
//
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public bool IsBalanced(TreeNode root) {
        return height(root) != -1;
    }
    
    private int height(TreeNode node) {
        if(node == null)
            return 0;
        int left = height(node.left);
        int right = height(node.right);
        if(left == -1 || right == -1 || Math.Abs(left - right) > 1)
            return -1;
        return Math.Max(left, right) + 1;
    }
}

//Third fastest
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool IsBalanced(TreeNode root) 
    {
        if (CheckLength(root) == -1) return false;
        return true;
    }
    
    private int CheckLength(TreeNode root)
    {
        if (root == null) return 0;
        
        int left = CheckLength(root.left);
        if (left == -1) return -1;
        
        int right = CheckLength(root.right);
        if (right == -1) return -1;
        
        int diff = Math.Abs(left - right);
        
        if (diff > 1) return -1;
        return 1 + Math.Max(left, right);
        
    }
}

//Second fastest
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool IsBalanced(TreeNode root) {
        return GetBalancedDepth(root) != -1;
    }
    
    private int GetBalancedDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        
        if (root.left == null && root.right == null) {
            return 1;
        }
        
        //if ((root.left == null) != (root.right == null)) {
            //var any = GetBalancedDepth(root.left ?? root.right);
            //return any == -1 ? -1 : any + 1;
            //return -1;
        //}
        
        var left = GetBalancedDepth(root.left);
        var right = GetBalancedDepth(root.right);
        
        if (left == -1 || right == -1 || Math.Abs(left - right) > 1) {
            return -1;
        }
        
        return Math.Max(left, right) + 1;
    }
}

//Fastest
**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool IsBalanced(TreeNode root)
        {
            if (root == null)
            {
                return true;
            }

            var heightOfLeft = Height(root.left);
            var heightOfRight = Height(root.right);

            if (Math.Abs(heightOfLeft - heightOfRight) <= 1
                && IsBalanced(root.left)
                && IsBalanced(root.right))
            {
                return true;
            }

            return false;
        }

        public int Height(TreeNode node)
        {
            if (node == null)
            {
                return 0;
            }

            var height = 1 + Math.Max(Height(node.left), Height(node.right));
            return height;
        }
}

File: Solutions/110-Balanced Binary Tree/110-Balanced Binary Tree.rb
#Third fastest
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}
def is_balanced(root)
    return true if root.nil?
    check_dept(root)[:status]
end

def check_dept(node)
    return { status: true, dept: -1 } unless node
    l_d = check_dept(node.left)
    r_d = check_dept(node.right)
    if (l_d[:dept] - r_d[:dept]).abs > 1 || !(l_d[:status] && r_d[:status])
        return { status: false, dept: [l_d[:dept], r_d[:dept]].max + 1  }
    end
    { status: true, dept: [l_d[:dept], r_d[:dept]].max + 1 }
end


#Second fastest
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}
def is_balanced(root)
  return true unless root  
  return false if height(root) == -1
  
  true
end

def height(node)
  return 0 unless node
  
  lh = height(node.left)
  rh = height(node.right)
  return -1 if lh == -1 || rh == -1
  return -1 if (lh - rh).abs > 1
  
  [lh, rh].max + 1
end


#Fastest
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}
def is_balanced(root)
  # return true if root.nil?
  # return false if (height(root.left)-height(root.right)).abs > 1
  # is_balanced(root.left) && is_balanced(root.right)
  helper(root)[0]
end

# [is_balanced, depth]
def helper(root)
  return [true, 0] if root.nil?
  left = helper(root.left)
  return [false, -1] unless left[0]
  right = helper(root.right)
  return [false, -1] unless right[0]
  return [false, -1] if (left[1] - right[1]).abs > 1
  [true, 1+[left[1], right[1]].max]
end

def height(root)
  return 0 if root.nil?
  1+[height(root.left), height(root.right)].max
end